### PR TITLE
Add FLOIP submission endpoint

### PR DESCRIPTION
**Pull request by nateMJ**
Closes https://github.com/onaio/onadata/issues/1118

### CODE REVIEW

File: onadata/apps/api/viewsets/xform_submission_viewset.py [hunk 1]
```diff
@@ -37,18 +42,33 @@ class XFormSubmissionViewSet(AuthenticateHeaderMixin,  # pylint: disable=R0901
...
                         BrowsableAPIRenderer)
     serializer_class = SubmissionSerializer
     template_name = 'submission.xml'
+    parser_classes = (FLOIPParser, JSONParser, FormParser, MultiPartParser)
+
+    def get_serializer(self, *args, **kwargs):
+        """
+        Pass many=True flag if data is a list.
+        """
+        if "data" in kwargs:
```
Comment (pld) [inline onadata/apps/api/viewsets/xform_submission_viewset.py]: could change lines 60-64 to be:

```python
data  = kwargs.get("data")

if isinstance(data, list):
    kwargs["many"] = True
```

File: onadata/libs/serializers/data_serializer.py [hunk 2]
```diff
@@ -303,3 +303,36 @@ def create(self, validated_data):
...
+    FLOIP SubmmissionSerializer - Handles a row of FLOIP specification format.
+    """
+    def to_internal_value(self, data):
+        """
+        Overrides validating rows in list data.
+        """
+        if not isinstance(data, list):
+            raise serializers.ValidationError(_(u'Invalid format. '
+                                                u'Expecting a list.'))
+        atleast_five = len(data) >= 5
```
Comment (pld) [inline onadata/libs/serializers/data_serializer.py]: doesn't look like it needs a var, can be made inline

Comment (pld) [reply onadata/libs/serializers/data_serializer.py]: can also remove the not

File: onadata/apps/api/tests/viewsets/test_xform_submission_viewset.py [hunk 3]
```diff
@@ -437,3 +436,62 @@ def test_head_submission_anonymous(self):
...
+            '/submission', data,
+            content_type='application/vnd.org.flowinterop.results+json')
+        response = self.view(request)
+        self.assertEqual(response.status_code, 401)
+        auth = DigestAuth('bob', 'bobbob')
+        request.META.update(auth(request.META, response))
+        response = self.view(request, username=self.user.username,
+                             xform_pk=self.xform.pk)
+        self.assertContains(response,
+                            "{u'non_field_errors': [u'Invalid format. "
```
Comment (pld) [inline onadata/apps/api/tests/viewsets/test_xform_submission_viewset.py]: This test isn't passing, not sure if the test is wrong or the `validate` method in the serializer

File: onadata/apps/api/viewsets/xform_submission_viewset.py [hunk 4]
```diff
@@ -60,6 +82,12 @@ def get_serializer_class(self):
         if 'application/x-www-form-urlencoded' in content_type:
             return RapidProSubmissionSerializer
 
+        if 'application/vnd.org.flowinterop.results+json' in content_type:
```
Comment (ukanga) [inline onadata/apps/api/viewsets/xform_submission_viewset.py]: Can we have this in a constant variable, it's being used in a couple of places in this file.

```python
FLOIP_RESULTS_CONTENT_TYPE = 'application/vnd.org.flowinterop.results+json'
```
Then
```python
if FLOIP_RESULTS_CONTENT_TYPE in content_type:
    ....
    request.accepted_media_type = FLOIP_RESULTS_CONTENT_TYPE
```

File: onadata/libs/serializers/data_serializer.py [hunk 5]
```diff
@@ -303,3 +303,48 @@ def create(self, validated_data):
...
+    """
+    def validate(self, attrs):
+        data = self.context['request'].data
+        error_msg = None
+
+        if not isinstance(data, list):
+            error_msg = u'Invalid format. Expecting a list.'
+        else:
+            for row_i in xrange(len(data)):
+                if len(data[row_i]) != 6:
```
Comment (ukanga) [inline onadata/libs/serializers/data_serializer.py]: Should `6` be a constant ?

File: onadata/libs/serializers/data_serializer.py [hunk 6]
```diff
@@ -303,3 +303,48 @@ def create(self, validated_data):
...
+                    error_msg = u"All rows must have 6 values, "\
+                                u"row %s does not." % row_i
+                    break
```
Comment (ukanga) [inline onadata/libs/serializers/data_serializer.py]: Could we state what the 6 values should be?

Also, you should probably apply `ugettext()` here for this to be available in translation files.
```python
_("All rows must have 6 values, row %(row)s does not." % {'row': row_i})
```
Applying on the variable should work when it is being translated but the string will not be included in the translation files.

File: onadata/libs/serializers/data_serializer.py [hunk 7]
```diff
@@ -303,3 +303,48 @@ def create(self, validated_data):
...
+        if error_msg:
+            raise serializers.ValidationError(_(error_msg))
+
+        return super(FLOIPSubmissionSerializer, self).validate(attrs)
+
+    def to_internal_value(self, data):
+        """
+        Overrides validating rows in list data.
+        """
+        data = {data[1]: data}
```
Comment (ukanga) [inline onadata/libs/serializers/data_serializer.py]: I feel like we should have a check first that data has at least one row and is probably not `None`.

File: onadata/libs/serializers/data_serializer.py [hunk 8]
```diff
@@ -303,3 +303,48 @@ def create(self, validated_data):
...
+        return data
+
+    def create(self, validated_data):
+        """
+        Returns object instances based on the validated data.
+        """
+        request, username = get_request_and_username(self.context)
+        xform_pk = self.context['view'].kwargs['xform_pk']
+        xform_id = get_object_or_404(XForm, pk=xform_pk).id_string
+        xform_headers = XForm.objects.get(id_string=xform_id).get_keys()
```
Comment (ukanga) [inline onadata/libs/serializers/data_serializer.py]: Why to queries here when we can do with only one?

Do note that `id_string` is not a unique field, it is possible to have the same id_string in different user accounts.

```python
xform = get_object_or_404(XForm, pk=xform_pk)
xform_headers = xform.get_keys()
```

File: onadata/libs/serializers/data_serializer.py (hunk 7 above)
Comment (nateMJ) [reply onadata/libs/serializers/data_serializer.py]: By this point, we get a row of data already. I can add the check at this point to confirm that.

```
elif len(data) > 0:
    for row_i in xrange(len(data)):
        if len(data[row_i]) != 6:
            error_msg = _(u"All rows must have 6 values, "\
                          u"row %(row)s does not." % {'row': row_i})
```

Comment (ukanga) [reply onadata/libs/serializers/data_serializer.py]: More likely,
```python
elif data:
    for row_i, row in enumerate(data):
        if len(row) != 6:
            error_msg = _(u"All rows must have 6 values, "\
                          u"row %(row)s does not." % {'row': row_i})
```

**Comment by pld** [comment]
Eesh, yeah that's so much better than my code

**Pull request merged**